// @google/gemini-cli/dist/src/utils/featureToggleUtils.js
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
import { SettingScope, isLoadableSettingScope, } from '../config/settings.js';
/**
 * Enables a feature by ensuring it is enabled in all writable scopes.
 */
export function enableFeature(settings, featureName, strategy) {
    const writableScopes = [SettingScope.Workspace, SettingScope.User];
    const foundInDisabledScopes = [];
    const alreadyEnabledScopes = [];
    for (const scope of writableScopes) {
        if (isLoadableSettingScope(scope)) {
            const scopePath = settings.forScope(scope).path;
            if (strategy.needsEnabling(settings, scope, featureName)) {
                foundInDisabledScopes.push({ scope, path: scopePath });
            }
            else {
                alreadyEnabledScopes.push({ scope, path: scopePath });
            }
        }
    }
    if (foundInDisabledScopes.length === 0) {
        return {
            status: 'no-op',
            featureName,
            action: 'enable',
            modifiedScopes: [],
            alreadyInStateScopes: alreadyEnabledScopes,
        };
    }
    const modifiedScopes = [];
    for (const { scope, path } of foundInDisabledScopes) {
        if (isLoadableSettingScope(scope)) {
            strategy.enable(settings, scope, featureName);
            modifiedScopes.push({ scope, path });
        }
    }
    return {
        status: 'success',
        featureName,
        action: 'enable',
        modifiedScopes,
        alreadyInStateScopes: alreadyEnabledScopes,
    };
}
/**
 * Disables a feature in the specified scope.
 */
export function disableFeature(settings, featureName, scope, strategy) {
    if (!isLoadableSettingScope(scope)) {
        return {
            status: 'error',
            featureName,
            action: 'disable',
            modifiedScopes: [],
            alreadyInStateScopes: [],
            error: `Invalid settings scope: ${scope}`,
        };
    }
    const scopePath = settings.forScope(scope).path;
    if (strategy.isExplicitlyDisabled(settings, scope, featureName)) {
        return {
            status: 'no-op',
            featureName,
            action: 'disable',
            modifiedScopes: [],
            alreadyInStateScopes: [{ scope, path: scopePath }],
        };
    }
    // Check if it's already disabled in the other writable scope
    const otherScope = scope === SettingScope.Workspace
        ? SettingScope.User
        : SettingScope.Workspace;
    const alreadyDisabledInOther = [];
    if (isLoadableSettingScope(otherScope)) {
        if (strategy.isExplicitlyDisabled(settings, otherScope, featureName)) {
            alreadyDisabledInOther.push({
                scope: otherScope,
                path: settings.forScope(otherScope).path,
            });
        }
    }
    strategy.disable(settings, scope, featureName);
    return {
        status: 'success',
        featureName,
        action: 'disable',
        modifiedScopes: [{ scope, path: scopePath }],
        alreadyInStateScopes: alreadyDisabledInOther,
    };
}
